render/egl.c

```diff
@@ -347,13 +369,113 @@ struct wlr_egl *wlr_egl_create(EGLenum platform, void *remote_display,
 		}
 	}
 
+	return true;
+}
+
+struct wlr_egl *wlr_egl_create(EGLenum platform, void *remote_display,
+		const EGLint *config_attribs) {
+	struct wlr_egl *egl = egl_create();
+	if (egl == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create EGL context");
+		goto error;
+	}
+
+	if (!egl_init(egl, platform, remote_display, config_attribs)) {
+		wlr_log(WLR_ERROR, "Failed to initialize EGL context");
+		goto error;
+	}
+
 	return egl;
 
 error:
 	eglMakeCurrent(EGL_NO_DISPLAY, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
 	if (egl->display) {
 		eglTerminate(egl->display);
 	}
+	if (egl) {
+		free(egl);
+	}
+	eglReleaseThread();
+	return NULL;
+}
+
+static bool device_has_name(const drmDevice *device, const char *name) {
+	for (size_t i = 0; i < DRM_NODE_MAX; i++) {
+		if (!(device->available_nodes & (1 << i))) {
+			continue;
+		}
+		if (strcmp(device->nodes[i], name) == 0) {
+			return true;
+		}
+	}
+	return false;
+}
+
+struct wlr_egl *wlr_egl_create_from_drm_fd(int drm_fd) {
+	struct wlr_egl *egl = egl_create();
+	if (egl == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create EGL context");
+		goto error;
+	}
+
+	if (egl->procs.eglQueryDevicesEXT == NULL) {
+		wlr_log(WLR_ERROR, "EGL_EXT_device_enumeration not supported");
+		goto error;
+	}
+
+	static const EGLint MAX_DEVICES = 64;
+	EGLint nb_devices = 0;
+	if (!egl->procs.eglQueryDevicesEXT(MAX_DEVICES, NULL, &nb_devices)) {
+		wlr_log(WLR_ERROR, "Failed to query EGL devices");
+		goto error;
+	}
+
+	EGLDeviceEXT *devices = calloc(nb_devices, sizeof(EGLDeviceEXT));
+	if (devices == NULL) {
+		wlr_log_errno(WLR_ERROR, "Failed to allocate device list");
+		goto error;
+	}
+
+	if (!egl->procs.eglQueryDevicesEXT(MAX_DEVICES, devices, &nb_devices)) {
+		wlr_log(WLR_ERROR, "Failed to query EGL devices");
+		goto error;
+	}
+
+	drmDevicePtr device = NULL;
+	drmGetDevice(drm_fd, &device);
+
+	EGLDeviceEXT *egl_device = NULL;
+	for(int i = 0; i < nb_devices; i++) {
+		const char *egl_device_name = egl->procs.eglQueryDeviceStringEXT(
+				devices[i], EGL_DRM_DEVICE_FILE_EXT);
+
+		if (device_has_name(device, egl_device_name)) {
+				wlr_log(WLR_DEBUG, "Using EGL device %s", egl_device_name);
+				egl_device = devices[i];
+				break;
+		}
+	}
+
+	if (egl_device == NULL) {
+		wlr_log(WLR_ERROR, "Failed to find EGLDeviceEXT");
+		goto error;
+	}
+
+	if (!egl_init(egl, EGL_PLATFORM_DEVICE_EXT, egl_device, NULL)) {
```

Comment:
Note: we should still pass an `EGLDeviceEXT *` here. Mesa will de-reference the pointer and expect an `EGLDeviceEXT`.